### PR TITLE
sync head

### PR DESCRIPTION
**Pull request by shayne-fletcher**
- Sync to https://gitlab.haskell.org/ghc/ghc.git `d7385f7077c6258c2a76ae51b4ea80f6fa9c7015`;
- Remove `happy-1.20.0` from extra-deps in hadrian stack.yaml now https://gitlab.haskell.org/ghc/ghc/-/commit/aaa51dcfdb729f130aeefeaeac15029b62096a74 has landed;
- Rework `applyPatchHadrianStackYaml` to be a little more resilient to the presence/absence of fields.

### CODE REVIEW

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 1]
```diff
@@ -579,19 +579,35 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
+   update fld elems =
+     writeFile hadrianStackYaml .
+       appendOrExtend fld elems
+       =<< readFile' hadrianStackYaml
+
+   -- If there are multiple occurences of a stanza (e.g. 'extra-deps')
+   -- they don't get merged and the last occurence prevails. Thus it's
+   -- important to extend a stanza if one already exists and not just
+   -- append a new one.
+   appendOrExtend :: String -> [String] -> String -> String
```
Comment (cocreature) [inline ghc-lib-gen/src/Ghclibgen.hs:604]: This seems pretty messy. Can we just use `yaml`/`aeson` instead and parse this properly?

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:604]: That seems total overkill for this. Further, I don't think that stack.yaml is either yaml or json. Look at the format of ghc-options vs extra-deps for an example.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:604]: well, that's wrong of course. presumably it's yaml.

Comment (cocreature) [reply ghc-lib-gen/src/Ghclibgen.hs:604]: I’m not sure I’m following. Why are those fields not valid yaml? Parsing yaml into an Aeson object is hardly and modifying the field is hardly any longer than this, much more robust and easier to understand.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:604]: yep. i'll see what i can do.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs]: thanks @cocreature i switched the code over to aeson/yaml as you suggested. let me know what you think!

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 2]
```diff
@@ -579,19 +589,42 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
 -- function 'appyPatchHadrianStackYaml' guarantees it is available
 -- if its needed.
 
--- Patch Hadrian's Cabal.
+data HadrianStackYaml =
+  HadrianStackYaml {
+      extraDeps :: [Y.Value]
+    , ghcOptions :: HMS.HashMap T.Text T.Text
+    , otherFields :: Y.Object
+  } deriving (Eq, Show)
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:597]: Can't we just do the modifications on the YAML value without this type and its instances? That should be significantly shorter and probably not harder to understand.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: I'm sorry, you'll have to demonstrate the technique. I don't know how to do that.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: By the way, what would be the goal? Is there something wrong with this program?

Comment (cocreature) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: parse into `Object` and do a key lookup on the hashmap instead of parsing into a custom type.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: Ah right, I get you. Well, the more verbose version exists. Is there real value in unwinding it in favor of doing that?

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: It is of course a short distance from here to there compared to from where we were.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs]: it's done

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 3]
```diff
@@ -579,19 +590,25 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
-       appendFile "hadrian/stack.yaml" $
-       unlines [
-           "extra-deps:"
-         , "  - happy-1.20.0" -- See https://gitlab.haskell.org/ghc/ghc/-/issues/18726.
-         , "  - exceptions-0.10.4" -- See [Note : GHC now depends on exceptions package]
-         ]
-     _ -> pure ()
+  let hadrianStackYaml = "hadrian/stack.yaml"
+  bytes <- B.readFile hadrianStackYaml
+  config <- Y.decodeThrow bytes
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:597]:
```suggestion
  config <- Y.decodeFileThrow hadrianStackYaml
```

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 4]
```diff
@@ -579,19 +590,25 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
+  -- See [Note : GHC now depends on exceptions package]
+  let deps = [x | ghcFlavor == GhcMaster, x <- [T.pack "exceptions-0.10.4"]] ++
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs]:
```suggestion
  let deps = ["exceptions-0.10.4" | ghcFlavor == GhcMaster] ++
```

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 5]
```diff
@@ -579,19 +590,25 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
+        case parse (\cfg -> cfg .:? "extra-deps" .!= []) config of
+          Success ls -> ls :: [T.Text]
+          _ -> []
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:602]: If `extra-deps` is not a list, something must have gone wrong. Let's not swallow the error but rather crash noisily.
```suggestion
          Error msg -> error msg
```

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 6]
```diff
@@ -579,19 +590,25 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
+ -- Build hadrian (and any artifacts we generate via hadrian e.g.
+ -- Parser.hs) as quickly as possible.
+  let opts = HMS.insert "$everything" "-O0 -j" $
+        case parse (\cfg -> cfg .:? "ghc-options" .!= HMS.empty) config of
+          Success os -> os :: HMS.HashMap T.Text Y.Value
+          _ -> HMS.empty
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:608]:
```suggestion
          Error msg -> error msg
```

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 7]
```diff
@@ -579,19 +589,25 @@ applyPatchCmmParseNoImplicitPrelude ghcFlavor =
...
-           "extra-deps:"
-         , "  - happy-1.20.0" -- See https://gitlab.haskell.org/ghc/ghc/-/issues/18726.
-         , "  - exceptions-0.10.4" -- See [Note : GHC now depends on exceptions package]
-         ]
-     _ -> pure ()
+  let hadrianStackYaml = "hadrian/stack.yaml"
+  bytes <- B.readFile hadrianStackYaml
+  config <- Y.decodeThrow bytes
+  -- See [Note : GHC now depends on exceptions package]
+  let deps = [x | ghcFlavor == GhcMaster, x <- [toJSON (T.pack "exceptions-0.10.4")]] ++
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:599]: `Y.Value` has an `IsString` instance and we don't need the binding:
```suggestion
  let deps = ["exceptions-0.10.4" | ghcFlavor == GhcMaster] ++
```

**Review by hurryabit — APPROVED** [review]
Much better. Thank you.

File: ghc-lib-gen/src/Ghclibgen.hs (hunk 3 above)
Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:597]: Ok

File: ghc-lib-gen/src/Ghclibgen.hs (hunk 4 above)
Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs]: Ok

File: ghc-lib-gen/src/Ghclibgen.hs (hunk 5 above)
Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:602]: Ok

File: ghc-lib-gen/src/Ghclibgen.hs (hunk 6 above)
Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:608]: ok

File: ghc-lib-gen/src/Ghclibgen.hs (hunk 7 above)
Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:599]: Ok

**Comment by shayne-fletcher** [comment]
Can we move this forward to `/azp run` please?

**Pull request merged**